sympy/geometry/polygon.py

```diff
@@ -164,42 +170,41 @@ def __new__(cls, *args, **kwargs):
         else:
             return Point(*vertices, **kwargs)
 
-        # reject polygons that have intersecting sides unless the
-        # intersection is a shared point or a generalized intersection.
-        # A self-intersecting polygon is easier to detect than a
-        # random set of segments since only those sides that are not
-        # part of the convex hull can possibly intersect with other
-        # sides of the polygon...but for now we use the n**2 algorithm
-        # and check if any side intersects with any preceding side,
-        # excluding the ones it is connected to
         try:
             convex = rv.is_convex()
         except ValueError:
             convex = True
-        if not convex:
-            sides = rv.sides
-            for i, si in enumerate(sides):
-                pts = si.args
+        if not convex and not ignore_int:
+            sides = {}
+            for i, side in enumerate(rv.directed_sides):
+                sides[i] = [side]
+            for i in sides:
```

Comment:
perhaps [it's time to add something like this](https://en.wikipedia.org/wiki/Bentley%E2%80%93Ottmann_algorithm)